presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java

```diff
@@ -268,6 +277,114 @@ private Object selectUniqueValue(String projection, Type expectedType, Session s
+        // * The test fails if retained size reaches 1MB.
+        // Note that 1MB is arbitrarily chosen and may be increased if a function implementation
+        // legitimately needs more.
+
+        long maxRetainedSize = 0;
+        int maxIterationCount = 0;
+        for (int iterationCount = 0; iterationCount < Math.max(1000, maxIterationCount * 4); iterationCount++) {
+            PageProcessorOutput output = processor.process(session.toConnectorSession(), SOURCE_PAGE);
+            // consume the iterator
+            Iterators.getOnlyElement(output);
+
+            long retainedSize = processor.getProjections().stream()
+                    .mapToLong(this::getRetainedSizeOfCachedInstance)
+                    .sum();
+            if (retainedSize > maxRetainedSize) {
+                maxRetainedSize = retainedSize;
+                maxIterationCount = iterationCount;
+            }
+
+            if (maxRetainedSize >= 1048576) {
+                fail(format("The retained size of cached instance of function invocation is likely unbounded: %s", projection));
+            }
+        }
+    }
+
+    private long getRetainedSizeOfCachedInstance(PageProjection projection)
+    {
+        Field[] fields = projection.getClass().getDeclaredFields();
+        long retainedSize = 0;
+        for (Field field : fields) {
+            field.setAccessible(true);
+            String fieldName = field.getName();
+            if (!fieldName.startsWith("__cachedInstance")) {
+                continue;
+            }
+            try {
+                retainedSize += getRetainedSizeOf(field.get(projection));
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return retainedSize;
+    }
+
+    private long getRetainedSizeOf(Object object)
+    {
+        Class type = object.getClass();
+        if (type == PageBuilder.class) {
+            return ((PageBuilder) object).getRetainedSizeInBytes();
+        }
+        else if (type.isArray()) {
+            if (type == int[].class) {
+                return sizeOf((int[]) object);
+            }
+            else if (type == boolean[].class) {
+                return sizeOf((boolean[]) object);
+            }
+            else if (type == byte[].class) {
+                return sizeOf((byte[]) object);
+            }
+            else if (type == long[].class) {
+                return sizeOf((long[]) object);
+            }
+            else if (type == short[].class) {
+                return sizeOf((short[]) object);
+            }
+            else if (type == Block[].class) {
+                Object[] objects = (Object[]) object;
+                return Arrays.stream(objects)
+                        .mapToLong(this::getRetainedSizeOf)
+                        .sum();
+            }
+            else {
+                throw new UnsupportedOperationException(format("Unknown type encountered: %s", type));
```

Comment:
Throwing `IllegalArgumentException` seems like a better fit here, what do you think?